### PR TITLE
Add support for inferring custom runtime parameters

### PR DESCRIPTION
**Pull request by petersalas**
This adds CLI support for inferring/setting the schema for custom runtime parameters. We use `typescript-json-schema` to infer the schema and set it in GraphQL. (This change does not add CLI support for setting default runtime parameters, but adds it to the underlying GraphQL mutation.)

The bulk of the change consists of refactoring `fixie serve` so that _either_ a tunnel URL change or parameter schema change triggers a new revision.

### CODE REVIEW

File: packages/fixie/src/agent.ts [hunk 1]
```diff
@@ -260,6 +262,39 @@ export class FixieAgent {
     return config as AgentConfig;
   }
 
+  private static inferRuntimeParametersSchema(agentPath: string): TJS.Definition | null {
+    // If there's a tsconfig.json file, try to use Typescript to produce a JSON schema
+    // with the runtime parameters for the agent.
+    const tsconfigPath = path.resolve(path.join(agentPath, 'tsconfig.json'));
+    if (!fs.existsSync(tsconfigPath)) {
+      term.yellow('⚠️ tsconfig.json not found. Your agent will not support runtime parameters.\n');
```
Comment (NickHeiner) [inline packages/fixie/src/agent.ts:270]: When giving a "not found" message, it's always nice to include the full path that was checked.

File: packages/fixie/src/agent.ts [hunk 2]
```diff
@@ -260,6 +262,39 @@ export class FixieAgent {
...
+    fs.writeFileSync(
+      tempPath,
+      `
+      import Handler from '${handlerPath}';
+      export type RuntimeParameters = Parameters<typeof Handler>[0] extends infer T ? T : {};
+      `
+    );
+    const program = TJS.programFromConfig(tsconfigPath, [tempPath]);
+    const schema = TJS.generateSchema(program, 'RuntimeParameters', settings);
+    if (schema && schema.type !== 'object') {
```
Comment (NickHeiner) [inline packages/fixie/src/agent.ts:291]:
```suggestion
    if (schema?.type !== 'object') {
```

File: packages/fixie/src/agent.ts [hunk 3]
```diff
@@ -684,8 +655,11 @@ export class FixieAgent {
 
     // The tunnel may yield different URLs over time. We need to create a new
     // agent revision each time.
-    for await (const currentUrl of deploymentUrlsIter) {
-      await waitForAgentToBeReady();
+    for await (const [currentUrl, runtimeParametersSchema] of this.zipAsyncIterables(
+      deploymentUrlsIter,
+      schemaGenerator
+    )) {
```
Comment (NickHeiner) [inline packages/fixie/src/agent.ts:661]: I feel like we should just be using event emitters but I don't think it's worth trying to change now.

**Review by NickHeiner — APPROVED** [review]
LGTM on an extremely quick review

File: packages/fixie/src/agent.ts (hunk 3 above)
Comment (petersalas) [reply packages/fixie/src/agent.ts:661]: Probably -- I briefly started down that path but (probably incorrectly) convinced myself it wasn't going to be significantly better.

File: packages/fixie/src/agent.ts (hunk 2 above)
Comment (petersalas) [reply packages/fixie/src/agent.ts:291]: Actually, this is intentional -- `null !== "object"`

**Pull request merged**